packages/zoe/tools/manualTimer.js

```diff
@@ -1,197 +1,97 @@
-import { makeScalarMapStore } from '@agoric/store';
 import { Far } from '@endo/marshal';
-import { makeNotifierKit } from '@agoric/notifier';
-import { makePromiseKit } from '@endo/promise-kit';
+import { buildManualTimer as build } from '@agoric/swingset-vat/tools/manual-timer.js';
 import { TimeMath } from '@agoric/swingset-vat/src/vats/timer/timeMath.js';
 
-import { eventLoopIteration } from './eventLoopIteration.js';
 import './types.js';
 import './internal-types.js';
 
-const { details: X } = assert;
+// we wrap SwingSet's buildManualTimer to accomodate the needs of
+// zoe's tests
+
+/**
+ * @typedef {{
+ *  timeStep?: RelativeTime,
+ *  eventLoopIteration?: () => Promise<unknown>,
+ * }} ZoeManualTimerOptions
+ */
+
+const nolog = (..._args) => {};
 
 /**
- * A fake clock that also logs progress.
  *
- * @param {(...args: any[]) => void} log
+ * A fake TimerService, for unit tests that do not use a real
+ * kernel. You can make time pass by calling `advanceTo(when)`, or one
+ * `timeStep` at a time by calling `tick()`.
+ *
+ * `advanceTo()` merely schedules a wakeup: the actual
+ * handlers (in the code under test) are invoked several turns
+ * later. Some zoe/etc tests want to poll for the consequences of
+ * those invocations. The best approach is to get an appropriate
+ * Promise from your code-under-test, wait for it to fire, and then
+ * poll. But some libraries do not offer this convenience, especially
+ * when they use internal "fire and forget" actions.
+ *
+ * To support those tests, the manual timer accepts a
+ * `eventLoopIteration` option. If provided, each call to `tick()`
+ * will wait for all triggered activity to complete before
+ * returning. That doesn't mean the `wake()` handler's result promise
+ * has fired; it just means there are no settled Promises still trying
+ * to execute their callbacks.
+ *
+ * The following will wait for all such Promise activity to finish
+ * before returning from `tick()`:
+ *
+ *  eventLoopIteration = () => new Promise(setImmediate);
+ *  mt = buildManualTimer(log, startTime, { eventLoopIteration })
+ *
+ * `tickN(count)` calls `tick()` multiple times, awaiting each one
+ *
+ * The first argument is called to log 'tick' events, which might help
+ * with "golden transcript" -style tests to distinguish tick
+ * boundaries
+ *
+ * @param {(...args: any[]) => void} [log]
  * @param {Timestamp} [startValue=0n]
- * @param {RelativeTime} [timeStep=1n]
+ * @param {ZoeManualTimerOptions} [options]
  * @returns {ManualTimer}
  */
-export default function buildManualTimer(log, startValue = 0n, timeStep = 1n) {
-  startValue = TimeMath.toAbs(startValue);
-  timeStep = TimeMath.toRel(timeStep);
-  let ticks = startValue;
 
-  /** @type {MapStore<Timestamp, ERef<TimerWaker>[]>} */
-  const schedule = makeScalarMapStore('Timestamp');
+const buildManualTimer = (log = nolog, startValue = 0n, options = {}) => {
+  const {
+    timeStep = 1n,
+    eventLoopIteration = () => 0,
```

Comment:
when I looked over the timer PR test change I didn’t see anything surprising except that eventLoopIteration has to be passed in as an arg. I remember skimming it and thinking “okay” but I just took another look and I’m confused why it has to be passed in as a function instead of a flag. E.g. `iterateEventLoop: boolean = false`.

@warner what motivates receiving an arbitrary function?

Reply:
There are a handful of "swingset" unit tests (as opposed to "unit" unit tests) which use the manual timer too, which means they import `manualTimer.js` into code that runs inside a real vat. And `setImmediate` is not available to vat code (the kernel doesn't populate it in the Compartments that houses the vat code) to keep vats from retaining agency after claiming their crank/delivery has finished. So `manualTimer.js` can't provide that functionality by itself.

We experimented with a version that used `if (globalThis.setImmediate)` to try and work with both, but it ended up being unsatisfying, and it wasn't clear that all tests would want (or could even tolerate) the automatic wait-for-flush behavior all the time, so it seemed better to give tests a choice.

That might be worth revisiting, now that we figured out the other test failures (which turned out to be the result of different assumption).

Reply:
Ah. I'll humbly ask that we make the common case simpler (without real vat) even if it makes the utility more complex. Consider:
- add default option `iterateEventLoop = false`
- make `eventLoopIteration` option default to undefined
- when `iterateEventLoop` is true, manualTimer does that automatically if it can
- if it can't, it uses `eventLoopIteration` option and throws an error if it's not set, informing the dev that with a real vat it must be passed in